nmt.go

```diff
@@ -0,0 +1,142 @@
+package nmt
+
+import (
+	"bytes"
+	"crypto"
+	"errors"
+	"fmt"
+
+	"github.com/liamsi/merkletree"
+)
+
+const (
+	LeafPrefix = 0
+	NodePrefix = 1
+)
+
+var (
+	ErrMismatchedNamespaceSize = errors.New("mismatching namespace sizes")
+)
+
+var _ merkletree.TreeHasher = &namespacedTreeHasher{}
+var _ Nmt = &NamespacedMerkleTree{}
+
+type NamespacedMerkleTree struct {
+	nidLen     int
+	baseHasher crypto.Hash
+	tree       *merkletree.Tree
+}
+
+func (n NamespacedMerkleTree) NamespaceSize() int {
+	return n.nidLen
+}
+
+func (n NamespacedMerkleTree) Push(data NamespacePrefixedData) error {
+	got, want := data.NamespaceSize(), n.NamespaceSize()
+	if got != want {
+		return fmt.Errorf("%w: got: %v, want: %v", ErrMismatchedNamespaceSize, got, want)
+	}
+	// TODO: or should we only push to the actual tree at the end
```

Comment:
In the LazyLedger prototype, the code for block production is responsible for making sure the data is pushed in order.

However, as a standalone library, it probably makes sense to handle this here and make it idiotproof.